File: dia_01/dia_01.py
import re
import os.path


#FILENAME = "/sample.txt"
#FILENAME = "/sample2.txt"
FILENAME = "/input.txt"
acum = 0


digit_map ={"zero":"0","one":"1","two":"2","three":"3","four":"4","five":"5","six":"6","seven":"7","eight":"8","nine":"9"}


def finddigit(linia, digit, current):
    pos = linia.find(digit)
    if(pos>=0 and pos<=current[1]):
        current[0]=digit
        current[1]=pos
    return current

def replace_digits(linia):
    res = finddigit(linia,"zero",["",99])
    res = finddigit(linia,"one",res)
    res = finddigit(linia,"two",res)
    res = finddigit(linia,"three",res)
    res = finddigit(linia,"four",res)
    res = finddigit(linia,"five",res)
    res = finddigit(linia,"six",res)
    res = finddigit(linia,"seven",res)
    res = finddigit(linia,"eight",res)
    res = finddigit(linia,"nine",res)
    if res[1]<99:
        linia = linia.replace(res[0],digit_map[res[0]])
        return replace_digits(linia)
    else:
        return linia
    

def extraurenumeros(linia):
    linia=re.sub("[^0-9]", "", linia)
    num_linia = linia[0] + linia[-1]
    return int(num_linia)

file1 = open(os.path.dirname(__file__) + FILENAME)
data = file1.readlines()

new_data = [[x if (x := "".join([v for k, v in digit_map.items() if line[i:].startswith(k)])) else line[i] for i in range(len(line))] for line in data]

acum=0
acum_meu=0
pos_llista =0
for linia in new_data:
   # print(linia)
    digits = [int(i) for i in linia if i.isdigit()]
    dig1=digits[0] * 10 + digits[-1]
    acum += digits[0] * 10 + digits[-1]
    linia_x = replace_digits(data[pos_llista])
    pos_llista+=1
    dig2=extraurenumeros(linia_x)
    acum_meu = acum_meu + extraurenumeros(linia_x)
    if dig1 != dig2:
        print(str(dig1) + " " + str(linia))
        print(str(dig2) + " " + linia_x)
        print(pos_llista)
print("num:" + str(acum))
print("num:" + str(acum_meu))

File: dia_01/dia_01.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

// #define FILENAME "sample.txt"
//#define FILENAME "sample2.txt"
#define FILENAME "input.txt"


int procesaLinea(char *linia, int len)
{
    int primer = -1;
    int ultim = -1;
    for (int t = 0; t < len; t++)
    {
        if (isdigit(linia[t]))
        {
            if (primer < 0)
            {
                primer = (int)linia[t] - 48;
            }
            ultim = (int)linia[t] - 48;
        }
    }
    printf("------%d ` %d = %d\n", primer, ultim, primer * 10 + ultim);
    return primer * 10 + ultim;
}

int main()
{

    FILE *fp;
    ssize_t read;

    char *linia = NULL;

    // tamany de la linia
    size_t len_read = 0;
    int len=0;

    // acumulat
    int acum = 0;

    // linia ambs els numeros passats a digits (exercisi B)
    char nova_linia[50];

    // obrim el fitxer en mode lectura
    fp = fopen(FILENAME, "r");
    if (fp == NULL)
        exit(EXIT_FAILURE); // constant definida a stdlib.h

    // anem llegint linia a linia fis que getline torna un -1

    while ((read = getline(&linia, &len_read, fp)) != -1)
    {

            printf("------------------------\n");
            printf("linia %s len:%d strlen:%d\n", linia,(int)len, (int)strlen(linia));
            len = (int)strlen(linia);
        // PROVA A: acum +=procesaLinea(linia,len);
        for (int t = 0; t < len; t++)
        {

            if (len - t > strlen("zero") && linia[t] == 'z' && linia[t + 1] == 'e' && linia[t + 2] == 'r' && linia[t + 3] == 'o')
                nova_linia[t] = '0';
            else if (len - t > strlen("one") && linia[t] == 'o' && linia[t + 1] == 'n' && linia[t + 2] == 'e')
                nova_linia[t] = '1';
            else if (len - t > strlen("two") && linia[t] == 't' && linia[t + 1] == 'w' && linia[t + 2] == 'o')
                nova_linia[t] = '2';
            else if (len - t > strlen("three") && linia[t] == 't' && linia[t + 1] == 'h' && linia[t + 2] == 'r' && linia[t + 3] == 'e' && linia[t + 4] == 'e')
                nova_linia[t] = '3';
            else if (len - t > strlen("four") && linia[t] == 'f' && linia[t + 1] == 'o' && linia[t + 2] == 'u' && linia[t + 3] == 'r')
                nova_linia[t] = '4';
            else if (len - t > strlen("five") && linia[t] == 'f' && linia[t + 1] == 'i' && linia[t + 2] == 'v' && linia[t + 3] == 'e')
                nova_linia[t] = '5';
            else if (len - t > strlen("six") && linia[t] == 's' && linia[t + 1] == 'i' && linia[t + 2] == 'x')
                nova_linia[t] = '6';
            else if (len - t > strlen("seven") && linia[t] == 's' && linia[t + 1] == 'e' && linia[t + 2] == 'v' && linia[t + 3] == 'e' && linia[t + 4] == 'n')
                nova_linia[t] = '7';
            else if (len - t > strlen("eight") && linia[t] == 'e' && linia[t + 1] == 'i' && linia[t + 2] == 'g' && linia[t + 3] == 'h' && linia[t + 4] == 't')
                nova_linia[t] = '8';
            else if (len - t > strlen("nine") && linia[t] == 'n' && linia[t + 1] == 'i' && linia[t + 2] == 'n' && linia[t + 3] == 'e')
                nova_linia[t] = '9';
            else
                nova_linia[t] = linia[t];
            printf("%c", nova_linia[t]);
        }
        acum += procesaLinea(nova_linia, len);
    }
    printf("acumulat es %d", acum);
    fclose(fp);
    if (linia)
        free(linia);
    exit(EXIT_SUCCESS);
}